Add Collector security documentation

tiffany76 opened the pull request:
This PR moves **end user** security documentation from a [README](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/security-best-practices.md) in the Collector core repository to the OTel docs website. 

Based on decisions in previous issues and PRs, the following are assumed:
- Security documentation intended for component developers will remain in the README, which will be amended in a follow-up PR.
- Both documents will cross-reference the other.
- The top-level Collector landing page and the Collector configuration page will link to the security documents.

Tracking issue: #3479
Related to: #3227

**NOTE:** Much of the work for this PR was done by @mjingle in #3652. With her permission, I am building on her efforts.

tiffany76 commented:
Still lots to do. I'll let everyone know when it's in a state fit for reviewing. :+1:

reyang commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,155 @@
+
+We recommend enabling only the minimum required components for a Collector
+configuration: one receiver and one exporter. Configuring only the minimum set
+of components exposes a minimum set of required ports.
```
"Ports" seem to be too specific, consider rephrasing it to "exposes a minimum attack surface" - it could be TCP/UDP ports, component dependency (SBOM), etc.

tiffany76 replied:
Changed in cbf67a9. Thanks!

reyang commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,155 @@
+## Receivers and exporters
+
+We recommend enabling only the minimum required components for a Collector
+configuration: one receiver and one exporter. Configuring only the minimum set
```
"one receiver and one exporter" seems very strict, what if the user wants to fork the data to two different targets? (e.g. are we suggesting that forking data is not recommended from security perspective?)

tiffany76 replied:
Thanks, all! The content so far has been copied from an earlier draft of this documentation. I think this sentence was an edit of the following statement in the core respository's [README](https://github.com/open-telemetry/opentelemetry-collector/blob/d6f568dd5b6ddfd87a7c6350545cb818f902a06f/docs/security-best-practices.md?plain=1#L63):

> The configuration drives the Collector's behavior and care should be taken to ensure the configuration only enables the minimum set of capabilities and as such exposes the minimum set of required ports.

I will remove the "one receiver and one exporter" restriction from this PR.

tiffany76 replied:
I've reworked this whole section (Receivers and exporters) based on the comments. Please let me know what you think. No rush. I still have lots to do. Thanks!

reyang commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,155 @@
+  settings using configuration parameters. If these settings are available, you
+  should proceed with caution before modifying the default configuration values.
+  Improperly setting these values might expose the OpenTelemetry Collector to
+  additional attack vectors including resource exhaustion.
```
Resource exhaustion is a great topic, I feel the current description is a bit misleading - even if the settings are correct, OpenTelemetry Collector can still cause resource exhaustion - is this the current state or there is already a solid solution for resource governance?

tiffany76 replied:
Hi @reyang, I removed the mention of resource exhaustion here. The topic is addressed in more detail in the Processors section. I'll let you know once I start work on it. Thanks!

tiffany76 replied:
Hi again. I updated the resource utilization section in 8605abb.

reyang commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,155 @@
+  otlp:
+    protocols:
+      grpc:
+        endpoint: localhost:4317
```
I don't recall the exact details, would this bind to both IPv4 and IPv6, or the Collector has a recommendation regarding how to specify IPv4/v6 explicitly?

tiffany76 replied:
I've rewritten the DOS safeguard section as well, with updated information about `localhost`. 

Rather than provide an example of binding to a pod's IP, I linked to the [README section](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/security-best-practices.md#safeguards-against-denial-of-service-attacks). Unless you think that content (Docker, Kubernetes, etc.) is relevant to end users and belongs on the website? I was trying to draw a line between end user and developer content, but I'm happy to make changes.

jpkrohling commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,155 @@
+
+To change the default endpoint to be `localhost`-bound in all components, enable
+the `component.UseLocalHostAsDefaultHost` feature gate. This feature gate will
+be enabled by default in the Collector in a future release.
```
We are in the future already!

jpkrohling commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,155 @@
+exporters. They are responsible for processing telemetry before it's analyzed.
+From a security perspective, processors are useful in a few ways.
+
+### Recommended processors
```
I don't think we should prescribe which processors people should use. We could provide some example configurations for common use-cases though.

tiffany76 replied:
Removed the section in favor of examples.

jpkrohling commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,155 @@
+
+<!-- start same page content in hosting-best-practices -->
+
+The `batch` and `memory_limiter` processors help ensure the OpenTelemetry
```
Given the discussions we are having around the batch processor, we should probably stop advertising it.

cc @mx-psi

jpkrohling replied:
After a second thought, perhaps we should stop talking about the memory limiter _processor_ as well, mentioning the _extension_ instead once it gets ready?

https://github.com/open-telemetry/opentelemetry-collector/tree/main/extension/memorylimiterextension

tiffany76 replied:
Based on @reyang's earlier comment that resource utilization is an important topic, it seems like we should say something about best practices here. If we don't recommend the batch and memory limiter processors, and the memory limiter extension is still in development, how should we advise users to safeguard resource utilization in their Collector configurations and hosting?

tiffany76 replied:
I took a shot at writing this section on my own (8605abb), but it could definitely use a developer's review. 

I mention batching and memory limiting, but I don't call out the processors or extensions specifically. Once the memory limiter extension is ready, we can update with an example.

mx-psi replied:
I think a general mention to batching and memory limiting should be more future-proof

jpkrohling commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,155 @@
+
+### Forwarding
+
+A forwarding extension is used when you need to collect telemetry that's not
```
I feel like this is misplaced here... I don't think the forwarding extension is related to security, nor is a best practice.

tiffany76 replied:
Removed in dbfd1b6. Thanks!

jpkrohling commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,155 @@
+while the `host_observer` requires the OpenTelemetry Collector to run in
+privileged mode.
+
+<!--- But what about Juraci's comment here: https://github.com/open-telemetry/opentelemetry.io/pull/3652/files?diff=unified&w=0#r1417409370 --->
```
It's OK -- what we shouldn't be telling people is to run things as root or blindly disable security protections (like selinux).

jpkrohling commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,155 @@
+### Subprocesses
+
+Extensions can also be used to run subprocesses when the Collector can't
+natively run the collection mechanisms (for example, FluentBit). Subprocesses
```
I'm also not sure where this is coming from: the only component that I know of spawning new processes is the jmx receiver.

tiffany76 replied:
This section comes directly from the [security README](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/security-best-practices.md#subprocesses). Should I remove it?

mx-psi replied:
I vote for removing this section

tiffany76 replied:
Removed in 961ac47.

jpkrohling commented on content/en/docs/security/hosting-best-practices.md:
```diff
@@ -0,0 +1,61 @@
+
+## Storing configuration information securely
+
+<!--- TODO: SHOULD ensure sensitive configuration information is stored securely. How? -->
```
Perhaps add a reference to the secrets management practices for Kubernetes?

tiffany76 replied:
I added this in 73f4c38.

jpkrohling commented on content/en/docs/security/hosting-best-practices.md:
```diff
@@ -0,0 +1,61 @@
+
+To limit the exposure of servers to authorized users:
+
+- Enable authentication, using bearer token authentication extensions and basic
```
```suggestion
- Enable authentication, for instance by using bearer token authentication extensions and basic
```

tiffany76 commented on content/en/docs/security/_index.md:
```diff
@@ -1,4 +1,28 @@
+Collector configuration. Running a secure Collector can help you
+
+- Protect telemetry that might contain sensitive information, such as personally
+  identifiable information (PII), application-specific data, or network traffic
```
Still searching for the existing guidance, but I've updated the wording as suggested. I'll keep looking for the PII reference.

tiffany76 commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,155 @@
+
+- Receivers and exporters can be either push- or pull-based. In either case, the
+  connection established should be over a secure and authenticated channel.
+- Unused receivers and exporters should be disabled to minimize the attack
```
Added a recommendation to use `ocb` and clarified that unused components should be removed from configurations. Thanks!

mx-psi commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,234 @@
+
+Some components can increase the security risk of your Collector pipelines.
+
+- Receivers and exporters can be push- or pull-based. In either case, you should
+  establish the connection at least over a secure channel, potentially
+  authenticated as well.
```
I don't think the push-pull thing matters much here. There are also other components you may care about properly securing (e.g. if you are using an https confmap provider or an extension):
```suggestion
- Receivers, exporters and other components should establish
  network connections over a secure channel, potentially
  authenticated as well.
```

mx-psi commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,234 @@
+instead of `0.0.0.0`. For more information, see
+[CWE-1327: Binding to an Unrestricted IP Address](https://cwe.mitre.org/data/definitions/1327.html).
+
+From Collector v0.110.0, the default endpoints for all servers in Collector
+components are set to `localhost:4317` for `gRPC` ports or `localhost:4318` for
+`http` ports. For earlier versions of the Collector, change the default endpoint
+from `0.0.0.0` to `localhost` in all components by enabling the
+`component.UseLocalHostAsDefaultHost`
+[feature gate](https://github.com/open-telemetry/opentelemetry-collector/tree/main/featuregate).
```
The `4317` and `4318` is specific to the OTLP receiver, not all components.
```suggestion
From Collector v0.110.0, the default host for all servers in Collector components is `localhost`. 
For earlier versions of the Collector, change the default endpoint
from `0.0.0.0` to `localhost` in all components by enabling the
`component.UseLocalHostAsDefaultHost`
[feature gate](https://github.com/open-telemetry/opentelemetry-collector/tree/main/featuregate).
```

mx-psi commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,234 @@
+
+<!--- TODO: Extensions SHOULD NOT expose sensitive health or telemetry data. How? What can you do? -->
+
+### Health and telemetry
+
+Extensions are available for health check information, Collector metrics and
+traces, and generating and collecting profiling data. When enabled with their
+default settings, all of these extensions, except the health check extension,
+are accessible only locally to the OpenTelemetry Collector. Take care to protect
+sensitive information when configuring these extensions for remote access, as
+they might expose it accidentally.
```
This is mentioned above already in a different fashion, I would just remove this section

tiffany76 replied:
Removed in 961ac47.

mx-psi commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,234 @@
+sensitive information when configuring these extensions for remote access, as
+they might expose it accidentally.
+
+### Collector's internal telemetry
+
+<!--- INSERT RECOMMENDATIONS HERE. For example:
+
+1. Remove zPages.
+1. Remove configuration endpoints.
+-->
```
I would remove this for now and add it back later (we can create an issue for this). I don't think we necessarily need this for the first version

tiffany76 replied:
Removed in 961ac47. I also created #5429 so we remember to iterate.

tiffany76 commented on content/en/docs/security/_index.md:
```diff
@@ -2,3 +2,41 @@
 weight: 970
 ---
+
+Learn how the OpenTelemetry project discloses vulnerabilities and responds to
+incidents. Find out how to ensure your observability data is collected and
+transmitted in a secure manner.
+
+## Common Vulnerabilities and Exposures (CVEs)
+
+For CVEs across all repositories, see
+[Common Vulnerabilities and Exposures](/docs/security/cve).
+
+## Incident response
+
+Learn how to report a vulnerability or find out how incident responses are
+handled in
+[Community incident response guidelines](/docs/security/security-response).
```
The trouble is that the links to the Collector security pages don't exist yet, until this PR is merged.

mx-psi replied:
Ah, alright!

tiffany76 commented on content/en/docs/security/config-best-practices.md:
```diff
@@ -0,0 +1,234 @@
+- Use the
+  [OpenTelemetry Collector Builder (`ocb`)](/docs/collector/custom-collector) to
+  create a Collector distribution that uses only the components you need.
+- If you find that you have unused receivers and exporters, remove them from
+  your configuration.
```
Sounds good to me.

tiffany76 commented:
Hi folks! I'm moving this PR out of draft. It's ready for review (finally)!

Just a couple thoughts:

- I think I have addressed all the comments so far, but I did rewrite most of the content, so you might have trouble finding your comments in the changed files. I'm sorry about that.
- I just reread @jpkrohling's [comment](https://github.com/open-telemetry/opentelemetry.io/pull/3652#issuecomment-1853643945) about keeping these docs closer to the Collector getting started docs. I know there was some back and forth about it. I don't have a preference either way and will move the files to the Collector docs, if that's what everyone agrees on.
- I will be AFK next week, so that will be a prime time for everyone to pile on the reviews. :smiley:

mx-psi commented on content/en/docs/security/hosting-best-practices.md:
```diff
@@ -0,0 +1,58 @@
+
+Use the Collector's own
+[internal telemetry](/docs/collector/internal-telemetry/) to monitor its
+performance. Scrape metrics from the Collector about its CPU, memory, and
```
I wouldn't use "scrape" here, since the metrics may be exported via OTLP (scrape makes it sound like we only support Prometheus for this, which is not really the case). Maybe just:

```suggestion
performance. Collect metrics from the Collector about its CPU, memory, and
```

svrnm commented on content/en/docs/security/_index.md:
```diff
@@ -2,3 +2,41 @@
 weight: 970
 ---
+
+Learn how the OpenTelemetry project discloses vulnerabilities and responds to
+incidents. Find out how to ensure your observability data is collected and
+transmitted in a secure manner.
```
should we move that into the frontmatter `description`, it reads like one

tiffany76 replied:
It might be a little long for the `description`. Let me take another shot at making the wording more intro-like.

tiffany76 replied:
I rewrote this in f099457.

svrnm reviewed (approved):
minor comment, but looks good to me overall, great work! 🎉

mx-psi commented on content/en/docs/security/hosting-best-practices.md:
```diff
@@ -0,0 +1,58 @@
+
+## Apply the rule of least privilege
+
+The Collector should not require privileged access, except where the data it's
+collecting is in a privileged location. For example, in order to get pod logs by
+mounting a node volume, the Collector daemonset needs enough privileges to get
+that data.
```
This section is worded differently from the others, I would rephrase it so that it more explicitly tells you what to do (use the least amount of privilege required).

tiffany76 replied:
Good catch on the change in wording! I used the example that @jpkrohling gave in a comment that I can't find now. I will work on rephrasing it to match the other examples on the page.

tiffany76 replied:
I rewrote the section in 3b86e29. Let me know what you think. Thanks!

mx-psi commented on content/en/docs/security/hosting-best-practices.md:
```diff
@@ -0,0 +1,58 @@
+If you are storing telemetry for processing, make sure to restrict access to
+those directories to prevent tampering with raw data.
+
+## Keep your secrets secret
```
To avoid the repetition maybe we can say this instead:
```suggestion
## Keep your secrets safe
```

tiffany76 commented:
/fix:refcache

tiffany76 commented:
@open-telemetry/sig-security-maintainers, @open-telemetry/collector-approvers, @open-telemetry/docs-approvers: Hi all, I'm asking for a last round of reviews. Please let me know if you'd like me to make any changes.

Also, please see my earlier comment and add your thoughts:

> I just reread @jpkrohling's [comment](https://github.com/open-telemetry/opentelemetry.io/pull/3652#issuecomment-1853643945) about keeping these docs closer to the Collector getting started docs. I know there was some back and forth about it. I don't have a preference either way and will move the files to the Collector docs, if that's what everyone agrees on.

mx-psi reviewed (approved):
This looks good to me for a first version

reyang commented on content/en/docs/security/hosting-best-practices.md:
```diff
@@ -0,0 +1,63 @@
+## Control access to server-like components
+
+Some Collector components such as receivers and exporters can function like
+servers. To limit access to authorized users, you should
```
```suggestion
servers. To limit access to authorized users, you should:
```

reyang reviewed (approved):
LGTM, well written!

tiffany76 commented:
We have an approval from each SIG, so I was thinking we could give it one more day and then merge this on Friday. How does that sound? 

@jpkrohling @open-telemetry/docs-approvers

reyang commented:
> We have an approval from each SIG, so I was thinking we could give it one more day and then merge this on Friday. How does that sound?
> 
> @jpkrohling @open-telemetry/docs-approvers

:shipit: 🚢

chalin reviewed (approved):
Overall looks good (did a quick scan, not an in depth read though).
I've added suggested fixes for section-local paths.
I'll let other maintainers give their formal approval for the content.

tiffany76 commented:
/fix:format
